hoodie-client/src/main/java/com/uber/hoodie/io/HoodieCommitArchiveLog.java

```diff
@@ -58,132 +70,192 @@ public HoodieCommitArchiveLog(HoodieWriteConfig config, FileSystem fs) {
+                        .withFs(fs)
+                        .overBaseCommit("").build();
+            } else {
+                return this.writer;
+            }
+        } catch(InterruptedException | IOException e) {
+            throw new HoodieException("Unable to initialize HoodieLogFormat writer", e);
         }
-        return success & deleteOtherInstants();
     }
 
-    private boolean deleteOtherInstants() {
-        // Delete clean and rollback files
-        List<HoodieInstant> toDelete = getInstantsToDelete().collect(Collectors.toList());
-        if(!toDelete.isEmpty()) {
-            log.info("Deleting actions " + toDelete);
-            return deleteInstants(toDelete);
+    private void close() {
+        try {
+            if(this.writer != null) {
+                this.writer.close();
+            }
+        } catch(IOException e) {
+            throw new HoodieException("Unable to close HoodieLogFormat writer", e);
         }
-        return true;
     }
 
-    private Stream<HoodieInstant> getInstantsToDelete() {
-
-        int maxCommitsToKeep = config.getMaxCommitsToKeep();
-        int minCommitsToKeep = config.getMinCommitsToKeep();
-
-        HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
-        HoodieTimeline cleanTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION));
-        if (!cleanTimeline.empty() && cleanTimeline.countInstants() > maxCommitsToKeep) {
-            // Actually do the commits
-            return cleanTimeline.getInstants().limit(cleanTimeline.countInstants() - minCommitsToKeep);
+    /**
+     * Check if commits need to be archived. If yes, archive commits.
+     */
+    public boolean archiveIfRequired() {
+        try {
+            List<HoodieInstant> instantsToArchive = getInstantsToArchive().collect(Collectors.toList());
+            boolean success = true;
+            if (instantsToArchive.iterator().hasNext()) {
+                this.writer = openWriter();
+                log.info("Archiving instants " + instantsToArchive);
+                archive(instantsToArchive);
+                success = deleteArchivedInstants(instantsToArchive);
+            } else {
+                log.info("No Instants to archive");
+            }
+            return success;
+        } finally {
+            close();
         }
-        return Stream.empty();
     }
 
-    private Stream<HoodieInstant> getCommitsToArchive() {
+    private Stream<HoodieInstant> getInstantsToArchive() {
 
         int maxCommitsToKeep = config.getMaxCommitsToKeep();
         int minCommitsToKeep = config.getMinCommitsToKeep();
 
         HoodieTable table = HoodieTable.getHoodieTable(new HoodieTableMetaClient(fs, config.getBasePath(), true), config);
+
+        HoodieTimeline cleanAndRollbackTimeline = table.getActiveTimeline().getTimelineOfActions(Sets.newHashSet(HoodieTimeline.CLEAN_ACTION,
```

Comment:
I left a comment just for this. The problem is to get the commit timeline, I would have to know which actions they comprise of and use them here, this will cause logic duplication say if anything were to change in the behavior of getCommitTimeline() then we would have the make the changes here as well. I think it's best to not duplicate it here. Thoughts ?